src/PasswordConstraintBase.php

```diff
@@ -0,0 +1,151 @@
+<?php
+
+namespace Drupal\password_enhancements;
+
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Plugin\ContainerFactoryPluginInterface;
+use Drupal\Core\Plugin\PluginBase;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Provides a base class for password constraints.
+ *
+ * @see \Drupal\password_enhancements\Annotation\PasswordConstraint
+ * @see \Drupal\password_enhancements\PasswordConstraintInterface
+ * @see \Drupal\password_enhancements\PasswordConstraintManager
+ * @see plugin_api
+ */
+abstract class PasswordConstraintBase extends PluginBase implements PasswordConstraintInterface, ContainerFactoryPluginInterface {
+
+  /**
+   * The password constraint ID.
+   *
+   * @var string
+   */
+  protected $uuid;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+
+    $this->setConfiguration($configuration);
```

Comment:
Calling this setter in a constructor triggers a warning for me especially because `parent::__construct()` saves the value properly (`$this->configuration = $configuration;`) and by calling ` $this->setConfiguration($configuration);` you either override the value that the parent set or you are doing a useless operation. 

Is there a problem with the structure of the incoming `$configuration` array? If yes then what and why it could occur? Do you only call the `setConfiguration()` just to ensure the `uuid` is not null rather an empty string when a new entity gets created?

```php
// Do something with the $configuration array if you must.
$this->uuid = $configuration['uuid'] ?? ''; // I guess the primary reason you are doing this because the UUID must be initialized and it should not be hidden from the constructor by calling a different property setter.
parent::__construct($configuration, $plugin_id, $plugin_definition);
````

Reply:
This is pretty much the same that `\Drupal\image\ImageEffectBase::__construct()` does. I can agree it could be done better, but if we're trying to achieve that, then we:
1) need to also fix core (in this regard), and/or
2) need to implement a completely custom solution (instead of relying on giants' shoulders, core code).

IOW: looks like the configuration needs to be stored on `$this->configuration` as-is, instead of having it stored in `$this->configuration['data']` (which would mean much deeper exported structures than needed, with quite an amount of custom code that's not in sync with core habits).

One might ask why do we have to deal with that UUID+data magic? The reason is twofold:
1) `\Drupal\password_enhancements\Form\PasswordConstraintFormBase::buildForm()` adds a subform for everything that's configurable on the plugin (ie. pretty much everything beyond the UUID). This is pretty much the same as `\Drupal\image\Form\ImageEffectFormBase::buildForm()`.
2) If we don't add that `data` subform in `\Drupal\password_enhancements\Form\PasswordConstraintFormBase::buildForm()` (and related methods), then we need to fix that everywhere (at least in our module) instead of relying on some LazyPluginCollection code (which can't handle that `data` properly by itself, as it shouldn't).

After getting rid of this `data` subform (and all the related jazz, including, but not limited to, this `setConfiguration()` call in the constructor), one question remains open: is there any reason core invented this `data` subform approach? The only reason I'm seeing for that ATM is that this `data` subform approach prevents the plugins from tampering with the whole form, as they'd only see their own part.

But OK, let's get rid of this `data` subform in our code, as it causes all this mess.